python/mcap-ros2-support/mcap_ros2/dynamic.py

```diff
@@ -0,0 +1,354 @@
+                parser_fn = FIELD_PARSERS[field.type.type]
+                if parser_fn is None:
+                    raise NotImplementedError(
+                        f"Parsing for type {field.type.type} is not implemented"
+                    )
+                value = parser_fn(reader)
+                setattr(msg, field.name, value)
+
+    return msg
+
+
+def generate_dynamic(schema_name: str, schema_text: str) -> Dict[str, DecoderFunction]:
+    """Generate a dynamic ROS2 message type from a schema text.
+
+    :param schema_name: The name of the schema defined in `schema_text`.
+    :param schema_text: The schema text to use for deserializing the message payload.
+    :return: A dictionary containing the generated message.
+    """
+    # Split schema_text by separator lines containing at least 3 = characters
+    # (e.g. "===") using a regular expression
+    msgdefs: Dict[str, MessageSpecification] = {
+        "builtin_interfaces/Time": TimeDefinition,
+        "builtin_interfaces/Duration": TimeDefinition,
+    }
+    generators: Dict[str, DecoderFunction] = {}
+    cur_schema_name = schema_name
+    for cur_schema_text in re.split(r"^={3,}$", schema_text, flags=re.MULTILINE):
+        cur_schema_text = cur_schema_text.strip()
+        if not cur_schema_text:
+            continue
+        # Check for a "MSG: pkg_name/msg_name" line
+        match = re.match(r"^MSG:\s+(\S+)$", cur_schema_text, flags=re.MULTILINE)
+        if match:
+            cur_schema_name = match.group(1)
+            # Remove this line from the message definition
+            cur_schema_text = re.sub(
+                r"^MSG:\s+(\S+)$", "", cur_schema_text, flags=re.MULTILINE
+            )
+
+        # Parse the package and message names from the schema name
+        # (e.g. "std_msgs/msg/String" -> "std_msgs")
+        pkg_name = cur_schema_name.split("/")[0]
+        msg_name = cur_schema_name.split("/")[-1]
+        short_name = (
+            pkg_name + "/" + msg_name
+        )  # Lookup by short name (e.g. "std_msg/String")
+        msgdef = parse_message_string(pkg_name, msg_name, cur_schema_text)
+
+        # Add the message definition to the dictionary
+        msgdefs[cur_schema_name] = msgdef
+        msgdefs[short_name] = msgdef
+
+        # Add the message generator to the dictionary
+        generator: DecoderFunction = make_read_message(cur_schema_name, msgdefs)
+        generators[cur_schema_name] = generator
+        generators[short_name] = generator
+
+    return generators
+
+
+def __repr__(self: Any) -> str:
+    fields = ", ".join(f"{field}={getattr(self, field)}" for field in self.__slots__)
+    return f"{self.__name__}({fields})"
+
+
+def __getstate__(self: Any):
+    """Support for Python pickling."""
+    return [getattr(self, x) for x in self.__slots__]
+
+
+def __setstate__(self: Any, state: Any):
+    """Support for Python pickling."""
+    for x, val in zip(self.__slots__, state):
+        setattr(self, x, val)
```

Comment:
when un-pickling, does the dynamically generated class have to be present in the program already? or is the whole class encoded somehow? could be nice to include an example of pickling